operator/internal/controller/pulumi/stack_controller.go

```diff
@@ -679,12 +680,27 @@ func (r *StackReconciler) Reconcile(ctx context.Context, request ctrl.Request) (
 	if resyncFreq.Seconds() < 60 {
 		resyncFreq = time.Duration(60) * time.Second
 	}
+
+	// If the update failed, allow it to be retried with an exponential
+	// cooldown. We start with a 5 minute cooldown and double that for each
+	// failed attempt, up to a max of 24 hours. Failed Updates are considered
+	// synced while inside this cooldown period.
+	cooldown := time.Duration(0)
+	if instance.Status.LastUpdate != nil && instance.Status.LastUpdate.State == shared.FailedStackStateMessage {
+		cooldown = 5 * time.Minute
+		cooldown *= time.Duration(math.Exp2(float64(instance.Status.LastUpdate.Attempts)))
+		cooldown = min(24*time.Hour, cooldown)
+	}
+
+	// more conditional logic around it being failed-but-too-soon-to-retry, so it descends into the "if synced" block.
 	synced := instance.Status.LastUpdate != nil &&
 		instance.Status.LastUpdate.Generation == instance.Generation &&
-		instance.Status.LastUpdate.State == shared.SucceededStackStateMessage &&
-		(isStackMarkedToBeDeleted ||
-			(instance.Status.LastUpdate.LastSuccessfulCommit == currentCommit &&
-				(!sess.stack.ContinueResyncOnCommitMatch || time.Since(instance.Status.LastUpdate.LastResyncTime.Time) < resyncFreq)))
+		((instance.Status.LastUpdate.State == shared.SucceededStackStateMessage &&
+			(isStackMarkedToBeDeleted ||
+				(instance.Status.LastUpdate.LastSuccessfulCommit == currentCommit &&
+					(!sess.stack.ContinueResyncOnCommitMatch || time.Since(instance.Status.LastUpdate.LastResyncTime.Time) < resyncFreq)))) ||
+			(!isStackMarkedToBeDeleted &&
+				instance.Status.LastUpdate.State == shared.FailedStackStateMessage && time.Since(instance.Status.LastUpdate.LastResyncTime.Time) < cooldown))
```

Comment:
This is really hairy! I'll probably break this out and put a table test around it.

Reply:
Maybe some inline funcs to give names to the sub-expressions?

Reply:
`!isStackMarkedToBeDeleted` was in here because we had a test that expected finalize to immediately retry a failed deletion, but we should do the same backoff in that case so I fixed the test.

I broke out `isSynced` into its own function to make it easier to unit test.